src/app/server/Mdns.cpp

```diff
@@ -78,37 +78,204 @@ chip::ByteSpan FillMAC(uint8_t (&mac)[8])
 
 } // namespace
 
-uint16_t gSecuredPort   = CHIP_PORT;
-uint16_t gUnsecuredPort = CHIP_UDC_PORT;
+#if CHIP_DEVICE_CONFIG_ENABLE_EXTENDED_DISCOVERY
+
+constexpr const char kExtendedDiscoveryTimeoutKeypairStorage[] = "ExtDiscKey";
+
+void MdnsServer::SetExtendedDiscoveryTimeoutSecs(int16_t secs)
+{
+    ChipLogDetail(Discovery, "SetExtendedDiscoveryTimeoutSecs %d", secs);
+    chip::DeviceLayer::PersistedStorage::KeyValueStoreMgr().Put(kExtendedDiscoveryTimeoutKeypairStorage, &secs, sizeof(secs));
+}
+
+int16_t MdnsServer::GetExtendedDiscoveryTimeoutSecs()
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    int16_t secs;
+
+    err = chip::DeviceLayer::PersistedStorage::KeyValueStoreMgr().Get(kExtendedDiscoveryTimeoutKeypairStorage, &secs, sizeof(secs));
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Discovery, "Failed to get extended timeout configuration err: %s", chip::ErrorStr(err));
+        secs = CHIP_DEVICE_CONFIG_EXTENDED_DISCOVERY_TIMEOUT_SECS;
+    }
+
+    ChipLogDetail(Discovery, "GetExtendedDiscoveryTimeoutSecs %d", secs);
+    return secs;
+}
+
+/// Callback from Extended Discovery Expiration timer
+void HandleExtendedDiscoveryExpiration(System::Layer * aSystemLayer, void * aAppState)
+{
+    MdnsServer::Instance().OnExtendedDiscoveryExpiration(aSystemLayer, aAppState);
+}
+
+/// Checks if extended discovery has expired and if so,
+/// stops commissionable node advertising
+/// Extended Discovery Expiration refers here to commissionable node advertising when NOT in commissioning mode
+void MdnsServer::OnExtendedDiscoveryExpiration(System::Layer * aSystemLayer, void * aAppState)
+{
+    if (mExtendedDiscoveryExpirationMs == TIMEOUT_CLEARED)
+    {
+        ChipLogDetail(Discovery, "HandleExtendedDiscoveryTimeout callback for cleared session");
+        return;
+    }
+    uint64_t now = mTimeSource.GetCurrentMonotonicTimeMs();
+    if (mExtendedDiscoveryExpirationMs > now)
+    {
+        ChipLogDetail(Discovery, "HandleExtendedDiscoveryTimeout callback for reset session");
+        return;
+    }
+
+    // reset advertising
+    CHIP_ERROR err;
+    err = chip::Mdns::ServiceAdvertiser::Instance().StopPublishDevice();
```

Comment:
Is there a reason we are forced to stop advertisement of all services just to remove one service (in this case, the commissionable node advertisement)?  I do see ChipMdnsPublishService that has no symmetric ChipMdnsUnpublishService method.  Is there something architecturally that forces us into this situation where we can write individual services, but have to delete them all just to remove one?

This seems inherently fragile.  For instance, what if the operational advertisement needs to change during commissionable node advertisement?  Won't that break this state machine that's operating through these timer callbacks?

It seems like this code would be far simpler if we could selectively publish and unpublish each service.

Reply:
We did have a plan to have Stop commands available for ex operational and commissionable and commissioner separately. So this is the plan, we're just not there yet. There are a couple of things that need to happen before that happens.
1) The platform mdns API needs to track all the services it actually started. It's coming, but it relies on a couple of other PRs that are also coming.
2) The platform impls actually need implementations for their individual stop commands. Right now they're all marked as not implemented.

The TL;DR is you're right and it's coming, but it's more work than it sounds like and this PR is working with what it has.

Reply:
Got it.  Thank you.